### DISCUSSION TITLE
Get ThemeResource textbox / richeditbox foreground and background color properties from code behind

### DISCUSSION THREAD

**Discussion by iomismo** [microsoft/microsoft-ui-xaml#7410] [Q&A]
Making a basic text editor with WinUI, MVVM. Stucked in default RichEditBox foreground and background color **viewmodel** properties based on selected theme, to show them in AppBar ColorPickerButtons (from CommunityToolkit, same results with "basic" ColorPicker).

Using ```Application.Current.Resources[]```:

Tried different resource names and still not getting correct colors in dark theme (white foreground and whatever background color uses RichEditBox). They always show light theme colors:

- ```TextControlForegroundFocused``` (that TextBox Class msdn page states it is "Text color when focused")
- ```TextFillColorPrimary```
- ```SystemControlBackgroundAltHighBrush```
- ```SystemColorButtonFaceColor```
- ```SystemColorButtonTextColor```

Same results using ```new TextBox().Resources``` as property value (TextBox because, again, msdn TextBox class page states that resources used by TextBox are shared with RichEditBox, among other controls. Same results using RichEditBox):

- ```TextBoxForegroundThemeBrush```
- ```TextBoxBackgroundThemeBrush```

So maybe the question is how to know what specific resource / themeresource names are used in specific control properties. I see resource names in resource xaml files are very descriptive but for now it is a mess for me. Also, I never see dark theme in those files, only Default, Light and HighContrast, and there are no colors matching dark theme, so it is evident that I am missing something (or many things :)

Actually, using in xaml works with selected theme, e.g:
```<controls:ColorPickerButton SelectedColor="{ThemeResource SystemBaseHighColor}"/>```

But not if using it in code:
```
<!--xaml-->
<controls:ColorPickerButton SelectedColor="{x:Bind ViewModel.SelectedForeColor, Mode=TwoWay}"/>

// cs
private Color _selectedForeColor = Application.Current.Resources["SystemBaseHighColor"];
```

It always shows white foreground color, despite of selected theme.

**Comment by jschwizer99** [comment]
Seems to be related to https://github.com/microsoft/microsoft-ui-xaml/issues/3783#issuecomment-801150950.

Meanwhile, I use the following code
```
// It shows one of the work-arounds for the missing DynamicResource issue.
// The problem is that winrt::Microsoft::UI::Xaml::Application::Current().Resources() itself does not update dynamically
// its theme resources. Thus, we need to search through the MergedDictionaries() to find the proper theme resource.
// This is an awful workaround just to get a proper resource value.
winrt::Microsoft::UI::Xaml::Media::Brush getThemeBrush(winrt::Microsoft::UI::Xaml::ElementTheme const& theme, winrt::hstring const& style)
{
	std::wstring themeString = (theme == Microsoft::UI::Xaml::ElementTheme::Dark) ? L"Dark" : L"Light";
	winrt::Windows::Foundation::IInspectable themeResource = nullptr;
	for (const auto& mergedDirectory : winrt::Microsoft::UI::Xaml::Application::Current().Resources().MergedDictionaries())
	{
		// Its hard to find a stable solution to identify the correct merged
		// directory as we cannot use HasKey(). Currently, we assume that we
		// have only "Light" and "Dark" custom themes. The first directory
		// has 3 themes.
		// See https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.resourcedictionary?view=winrt-19041
		if (mergedDirectory.ThemeDictionaries().Size() == 2)
		{
			themeResource = mergedDirectory.ThemeDictionaries().TryLookup(winrt::box_value(themeString));
			break;
		}
	}
	if (themeResource)
	{
		winrt::Windows::Foundation::IInspectable color = themeResource.as<winrt::Microsoft::UI::Xaml::ResourceDictionary>().TryLookup(winrt::box_value(style));
		if (color)
		{
			auto c = winrt::unbox_value<winrt::Microsoft::UI::Xaml::Media::SolidColorBrush>(color).Color();
			return winrt::Microsoft::UI::Xaml::Media::Brush(
				winrt::unbox_value<winrt::Microsoft::UI::Xaml::Media::SolidColorBrush>(color));
		}
	}
	return nullptr;
}
```
